prow/tide/tide_test.go

```diff
@@ -1423,3 +1493,210 @@ func TestSync(t *testing.T) {
+					},
+				},
+				{
+					Name: "never",
+				},
+			},
+			expectedPresubmits: map[int]sets.String{100: sets.NewString("presubmit")},
+		},
+		{
+			name: "doesn't run against branch",
+			presubmits: []config.Presubmit{
+				{
+					Name:      "presubmit",
+					AlwaysRun: true,
+					Brancher: config.Brancher{
+						Branches: []string{"release", "dev"},
+					},
+				},
+				{
+					Name:      "always",
+					AlwaysRun: true,
+				},
+				{
+					Name: "never",
+				},
+			},
+			expectedPresubmits: map[int]sets.String{100: sets.NewString("always")},
+		},
+		{
+			name: "run_if_changed (uncached)",
+			presubmits: []config.Presubmit{
+				{
+					Name:         "presubmit",
+					RunIfChanged: "^CHANGE.$",
+				},
+				{
+					Name:      "always",
+					AlwaysRun: true,
+				},
+				{
+					Name: "never",
+				},
+			},
+			expectedPresubmits:  map[int]sets.String{100: sets.NewString("presubmit", "always")},
+			expectedChangeCache: map[string][]string{"org/repo#100:sha": {"CHANGED"}},
+		},
+		{
+			name: "run_if_changed (cached)",
+			presubmits: []config.Presubmit{
+				{
+					Name:         "presubmit",
+					RunIfChanged: "^FIL.$",
+				},
+				{
+					Name:      "always",
+					AlwaysRun: true,
+				},
+				{
+					Name: "never",
+				},
+			},
+			initialChangeCache:  map[string][]string{"org/repo#100:sha": {"FILE"}},
+			expectedPresubmits:  map[int]sets.String{100: sets.NewString("presubmit", "always")},
+			expectedChangeCache: map[string][]string{"org/repo#100:sha": {"FILE"}},
+		},
+		{
+			name: "run_if_changed (cached) (skippable)",
+			presubmits: []config.Presubmit{
+				{
+					Name:         "presubmit",
+					RunIfChanged: "^CHANGE.$",
+				},
+				{
+					Name:      "always",
+					AlwaysRun: true,
+				},
+				{
+					Name: "never",
+				},
+			},
+			initialChangeCache:  map[string][]string{"org/repo#100:sha": {"FILE"}},
+			expectedPresubmits:  map[int]sets.String{100: sets.NewString("always")},
+			expectedChangeCache: map[string][]string{"org/repo#100:sha": {"FILE"}},
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Logf("Starting test case: %q", tc.name)
+
+		if tc.initialChangeCache == nil {
+			tc.initialChangeCache = map[string][]string{}
+		}
+		if tc.expectedChangeCache == nil {
+			tc.expectedChangeCache = map[string][]string{}
+		}
+
+		cfg := &config.Config{}
+		cfg.SetPresubmits(map[string][]config.Presubmit{
+			"org/repo": tc.presubmits,
+			"foo/bar":  {{Name: "wrong-repo", AlwaysRun: true}},
+		})
+		cfgAgent := &config.Agent{}
+		cfgAgent.Set(cfg)
+		sp := subpool{
+			org:    "org",
+			repo:   "repo",
+			branch: "master",
+			prs:    []PullRequest{samplePR},
+		}
+		c := &Controller{
+			ca:               cfgAgent,
+			fileChangesCache: tc.initialChangeCache,
+			ghc:              &fgc{},
+		}
+
+		presubmits, err := c.presubmitsByPull(sp)
+		if err != nil {
+			t.Fatalf("unexpected error from presubmitsByPull: %v", err)
+		}
+		if !reflect.DeepEqual(presubmits, tc.expectedPresubmits) {
```

Comment:
These work now because you only keep a single entry in the map in all of your tests. If somebody adds a test with more than one entry, it will be flaky. Use `k8s.io/apimachinery/pkg/api/equality/Semantic.DeepEqual`

Reply:
Why would multiple elements in the map cause flakiness? We use `reflect.DeepEqual` to compare maps in a number of Prow tests. https://golang.org/src/reflect/deepequal.go#L149

Reply:
I was under the impression this wasn't working with reflect's DeepEqual.

/lgtm

then